### ISSUE TITLE
Is pydantic the right choice (mongoengine issue)

### ISSUE THREAD

**Issue by jwag956** [duo-labs/py_webauthn#110]
This library is middleware and will be used in who knows how many projects with lots of layers above and below it. Pydantic isn't very pythonic in what it allows. I am testing flask-security as a RP using mongoengine as a backend. It implements its BinaryField as a clever subclass of 'bytes' - but pydantic doesn't understand it.

Here is a small repro:
```
from pydantic import BaseModel

class Test1(BaseModel):
    id: bytes

class Binary(bytes):
    def __new__(cls, data):
        self = bytes.__new__(cls, memoryview(data).tobytes())
        return self

A = Binary(b"hello")
T = Test1(id=A)
```

The Binary class is from mongoengine using the bson package. If you run this you get:
```
Traceback (most recent call last):
  File "/Users/jwag/fs/flask-security/flask_security/pydantic_test.py", line 17, in <module>
    T = Test1(id=A)
  File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Test1
id
  Expected bytes, got Binary (type=type_error)
```

Pydantic doesn't allow subclasses (very un-pythonic IMHO) - and the only things I solution I found on the web was to write custom pydantic validators that check for isinstance()...

Other 'create typed data structures easily' methods such as builtin dataclasses and attr don't have this issue (they don't validate by default).

Ultimately I suppose I am asking - should a middleware package require validation? or let the developer decide - by using mypy etc..

**Comment by MasterKale** [comment]
Hmm, interesting problem. I see [pydantic supports the use of `typing.Type`](https://pydantic-docs.helpmanual.io/usage/types/#type) to:

> ...specify that a field may only accept classes (not instances) that are subclasses of `T`

What happens if you define `id` as `Type[bytes]`? Can you pass in `Binary` then?

I'd never considered this before. On the one hand I think there's a lot of value to providing run-time validation of WebAuthn responses as a first line of defense against malformed responses. On the other hand I can see how this might be undesirable as it's essentially imposing dependency restrictions on projects that want to use this library.

I've never used mongoengine before, and its docs are a little on the light side: does `BinaryField` not offer a way to return its internal value as `bytes`?

**Comment by jwag956** [comment]
a. I thought the same thing - Type[bytes] - but that actually means you need to pass in a class - not an instance!

b. mongoengine - actually pymongo - look at the Binary class here: https://github.com/mongodb/mongo-python-driver/blob/master/bson/binary.py  The comment in there says:
In python 3 instances of Binary with subtype 0 will be decoded
      directly to :class:`bytes`.

Which they are!

Python of course is dynamically / duck-typed. If someone passes in something that behaves like bytes -I believe the pythonic thing to do is accept and use it! If it doesn't behave like bytes - python will throw a runtime error.

Type hints are a great improvement to python IMHO - as it them allows developers to decide how much/strict static or run time checking they want/need.

Here is an interesting (though somewhat biased!) discussion around pydantic, attrs, dataclasses...:
https://threeofwands.com/why-i-use-attrs-instead-of-pydantic/

I did get it to work using a custom validator:

```
from pydantic import BaseModel, validator


class Test1(BaseModel):
    id: bytes

    @validator("id", pre=True)
    def subclasses_ok(cls, v):
        if not isinstance(v, bytes):
            raise ValueError("Must be subclass of bytes")
        return bytes(v)


class Binary(bytes):
    def __new__(cls, data: bytes):
        self = bytes.__new__(cls, memoryview(data).tobytes())
        return self


A = Binary(b"hello")
print(isinstance(A, bytes))

T = Test1(id=A)
```

**Comment by MasterKale** [comment]
@jwag956 Take a look at PR #111 and let me know what you think. There's a new **tests/test_bytes_subclass_support.py** I added to try and capture this use case and ensure future compatibility.

**Comment by MasterKale** [comment]
This issue should be resolved with the newly-released [py_webauthn==1.2.0](https://pypi.org/project/webauthn/)
